allennlp/training/log_writer.py

```diff
@@ -1,284 +1,351 @@
-from typing import Any, Callable, Dict, List, Optional, Set
+from collections import deque, defaultdict
+from typing import Any, Dict, List, Optional, Set, Union, Deque
 import logging
 
 import torch
 
-from allennlp.common.from_params import FromParams
+from allennlp.common import Registrable
 from allennlp.data import TensorDict
 from allennlp.nn import util as nn_util
 from allennlp.training.optimizers import Optimizer
 from allennlp.training import util as training_util
 from allennlp.models.model import Model
 
+
 logger = logging.getLogger(__name__)
 
 
-class LogWriter(FromParams):
+class LogWriter(Registrable):
     """
     Class that handles trainer logging.
 
     # Parameters
 
-    serialization_dir : `str`, optional (default = `None`)
-        If provided, this is where the logs will be written.
+    serialization_dir : `str`
+        The training serialization directory.
+
+        In a typical AllenNLP configuration file, this parameter does not get an entry in the
+        file, it gets passed in separately.
+
+    model : `Model`
+        The model being trained.
+
+        In a typical AllenNLP configuration file, this parameter does not get an entry in the
+        file, it gets passed in separately.
+
+    optimizer : `Optimizer`
+        The optimizer used to train the model.
 
-        In a typical AllenNLP configuration file, this parameter does not get an entry under the
-        "tensorboard_writer", it gets passed in separately.
+        In a typical AllenNLP configuration file, this parameter does not get an entry in the
+        file, it gets passed in separately.
 
     summary_interval : `int`, optional (default = `100`)
         Most statistics will be written out only every this many batches.
 
     distribution_interval : `int`, optional (default = `None`)
```

Comment:
We should really have a blog post about this some time too. This is the kind of thing that's super annoying to do yourself.